### ISSUE TITLE
ERC1155: allow acceptance check to be extendable

### ISSUE THREAD

**Issue by Zergity** [OpenZeppelin/openzeppelin-contracts#4252]
**🧐 Motivation**
User token should be able to override and customize the acceptance check functions

**📝 Details**
Change the scope of the following 2 functions from `private` to `internal virtual`:
* _doSafeTransferAcceptanceCheck
* _doSafeBatchTransferAcceptanceCheck

**Comment by Amxx** [comment]
Hello @Zergity 

> User (...) should be able to (...) customize the acceptance check functions

We always felts the opposite way. Do you have any example of overrides that you think would be desirable ?

**Comment by Zergity** [comment]
> Hello @Zergity
> 
> > User (...) should be able to (...) customize the acceptance check functions
> 
> We always felts the opposite way. Do you have any example of overrides that you think would be desirable ?

We're building an 1155-To-20-Shadow token where each 1155 ID is able to deploy (on-demand) a Shadow ERC20 token that shares control with the original 1155 ID but with a separate address for composability. When this Shadow ERC20 is transferred, the acceptance check will be bypassed, as almost no 3rd party contracts implement any acceptance check for ERC20.

Shadow tokens are useful because more and more projects want to use ERC1155 for their multiple tokens design but also want each of the token IDs to be composability with other DeFi applications. ERC20 wrapping is inconvenient for users, but it is also impractical to clone all IDs to ERC20 on launch. The most effective way to do this is to use the ERC1155 as the application's main design; then allow anyone to deploy a Shadow token that shares control with the original ID on their demand.

An EIP is planned to submit for this. And this PR can help this EIP simplify the Implementation by simply extending and overriding the OZ's implementation.

**Comment by Amxx** [comment]
> We're building an 1155-To-20-Shadow token where each 1155 ID is able to deploy (on-demand) a Shadow ERC20 token that shares control with the original 1155 ID but with a separate address for composability. When this Shadow ERC20 is transferred, the acceptance check will be bypassed, as almost no 3rd party contracts implement any acceptance check for ERC20.

That sounds like [something I've prototypes a few months ago](https://gist.github.com/Amxx/397908c7a6f6c5afcc673577d1901155).

Its interesting to notice that indeed, if you try to transfer the ERC20 version of the token, the underlying "ERC1155.safeTransferFrom" will require the target to implement the receiver hook. That makes it incompatible with Uniswap and other Defi protocoles.

My concern is that is you disable or alter the acceptance check, you would be breaking the ERC1155 [safe transfer rules](https://eips.ethereum.org/EIPS/eip-1155#safe-transfer-rules), and as a consequence your token would not really be ERC-1155 compliant. That is why the function is private: to make sure devs don't (accidentally) break the ERC-1155 compliance.

I'd love to have @frangio's and @ernestognw's opinion on that, because I'm personally puzzled.

**Comment by Zergity** [comment]
> That sounds like [something I've prototypes a few months ago](https://gist.github.com/Amxx/397908c7a6f6c5afcc673577d1901155).

Yes, that's exactly what we're trying to do. We have a full implementation in development here if you want to take a look: https://github.com/derivable-labs/shadow-token/tree/dev/contracts

**Comment by frangio** [comment]
On one hand I also believe that this customization would not be compliant with ERC-1155, unless you somehow had different rules for the ERC20 shadow token transfers than you do for the standard ERC-1155 transfers... in which case it starts to sound similar to ERC-777's compatibility with ERC-20 and I don't think that was a good idea.

On the other hand, I can't find good reasons to _prevent_ you from implementing this if you want to, so it might be OK to make the acceptance check internal and overrideable... @Amxx can you see reasons against making it internal?

**Comment by Amxx** [comment]
I think we agree,

I'm not really comfortable with the construction @Zergity shared, because it breaks many assumption one might make about the behavior of the shadow (event not emitted, approval being all or nothing, ...). But should be prevent devs from doing that? I think it would do more harm then good.

I'd rather have @Zergity break the ERC requirement knowingly, using our code (from npm) under the hood, than him doing the same with our contract "vendored" and "modified" directly from github.

**Comment by Zergity** [comment]
> I'm not really comfortable with the construction @Zergity shared, because it breaks many assumption one might make about the behavior of the shadow (event not emitted, approval being all or nothing, ...). 

Thank you for your comments. I'm not sure what did I missed, but all ERC20 events are emitted from the Shadow.

And the uint approval is planned to be implemented, as the implementation is still WIP.

**Comment by Amxx** [comment]
> all ERC20 events are emitted from the Shadow.
They are emitted if someone triggers the transfer on the shadow itself. 
But if the transfer is triggered using the ERC1155 interface on the factory, the shadow will not emit the ERC20 events corresponding to the balance change.

**Comment by Zergity** [comment]
> > all ERC20 events are emitted from the Shadow.
> 
> They are emitted if someone triggers the transfer on the shadow itself. But if the transfer is triggered using the ERC1155 interface on the factory, the shadow will not emit the ERC20 events corresponding to the balance change.

I've put a lot of thought into this because checking whether an ID has a Shadow requires querying untouched contracts and storage. Per design, there's only a small portion of 1155 IDs will have shadows, not the majority of them, so the trade-off must be carefully considered. Even if the ERC-20 events are emitted for 1155 transfers, past events (before the Shadow is deployed) are completely missing.

That's why I decide that 1155 transfers are an internal mechanism of the Shadow token, not the ERC-20 behavior itself. Some token constantly inflates/deflates or are manually rebased, which modify the user balances or moves fund around and still be ERC-20 compatible.

**Comment by Zergity** [comment]
> Yes, you'd need to ensure shadows are deployed for any tokenId that exist in the erc1155 contract, and all transfers emit events on both side from the start.

So that's where our ideas are different. Our Shadow standard is designed for applications where only a small number of IDs need an ERC20, on-demand, and only long after the protocol launch.
